### ISSUE TITLE
tB considers all Dim statements procedure-level and ignores Option Explicit

### ISSUE THREAD

**Issue by fafalone** [twinbasic/twinbasic#1606]
**Describe the bug**
This is a major departure in behavior from VB6; in VB6, if you have Option Explicit on, the Dim statement for a variable must appear prior to it. tB further allows violating Option Explicit in the case the Dim statement appears within an `If` or `Select` block that evaluates to `False`.

**To Reproduce**
1) New StdExe, Option Explicit enabled.

2) 
```
Class Form1
Option Explicit
    
    Private Sub Form_Load() Handles Form.Load
    x = 1
    If InStr(App.Path, "v") Then
        Dim x As Long
        MsgBox x
    End If
    MsgBox x
    End Sub
    
End Class
```
3) If the conditional is `True` for you, pick a string where it's false to see x is still defined; or if it's `False`, one that will make it `True` to see that. `If [True/False] Then` appears to have the issue too but I wanted to try something that couldn't be resolved before runtime.

**Expected behavior**
As in VB6, an error should occur on x = 1. 

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 10
 - twinBASIC compiler version 0.15.347

**Additional context**

**Comment by bclothier** [comment]
While I agree this is a departure, I should point out that the branches aren't relevant. This is valid:

```
Public Sub TestIt()
    If False Then
        Dim x As Long
    End If
    
    Debug.Print x
End Sub
```

because the first use of the variable is still after the `Dim` statement, even though technically it is in a block of unreachable code. The real problem is that there's a use of the variable before the `Dim` statement, and the presence of branching or absence thereof is not relevant.

**Comment by WaynePhillipsEA** [comment]
I agree... I'm not seeing how it's violating `Option Explicit`.  Yes, in tB you can currently `Dim` a variable at any point ... it's a feature ;)  I'm certain there will be a trend of everyone grouping their variable declarations at the bottom of their procedures.

**Comment by bclothier** [comment]
I think this may be a bigger problem:
```
    Private Sub Derp()
        Debug.Print TypeName(x)
        ReDim x(0) As Long
        Debug.Print TypeName(x)
        Dim x As String
    End Sub
```

In VBA, that's at least 3 compile errors (`Variable not defined`,  `Expected array`,  `Can't change data types of array elements`).  In tB, it crashes with access violation. 

While I don't think people will rush to move their variable declarations to the bottom of procedure, this can cause confusion if they are in habit of declaring the variable in the middle of the body but unintentionally uses it earlier. 

For the record, I never really liked the fact that VBx's `Dim` statement don't really have any scope smaller than procedure yet acts as if it was imperative. Very inconsistent but c'est la vie.

**Comment by bclothier** [comment]
Thank you!

For the sake of readability, I think I'd prefer at least a warning regarding first use of variable prior to its `Dim` statement. That way, it can be quickly fixed to put the `Dim` preceding the first use. The scoping issue is probably best discussed in the #1463, I think.

**Comment by WaynePhillipsEA** [comment]
Sorry, I was just jesting.   I don't really see any benefit to allowing the use of a variable before its definition, so I think that should be an compile error to match VBx, though not high priority at the moment.

**Comment by bclothier** [comment]
No worries, my humor detector is at a shop, so... 😁 

I did imagine that it would be a PITA to detect the first use before declaration (Rubberduck had plenty issues with the particular inspection in this area). As I said before, `Dim` statement is one special snowflake. 

Agreed RE: priority.

**Comment by fafalone** [comment]
I considered it violating the idea of Option Explicit because at the time the variable is used, you wouldn't expect it to exist. 

You wouldn't expect it to exist outside a conditional that's false alas I guess we're stuck with that one.

**Comment by WaynePhillipsEA** [comment]
Yeah, understood, I can see how you came to that.   This demonstrates the behaviour more clearly:
```
MsgBox TypeName(x) 
Dim x As Long
```
Shows "Long".  If it was treated as an implicit variable (Option Explicit off) then it would of course be "Variant".

**Comment by Tragen** [comment]
I would love to see future improvements where you can tell the compiler to be more strict with scopes but for VB6 compatibility we need it like it is now, even if it's stupid design by MS.

**Comment by mansellan** [comment]
I really like what .Net did when it retrofitted nullable reference types into its type system.

If the compiler can't _guarantee_ that a variable is non-null before it's first accessed, it offers an IDE hint that you may be in for some runtime explosions. It's a hint because it's impossible to know for sure at compile-time that a block is unreachable, but it is possible some of the time to prove that a variable will be initialised, and make the hint goes away. It's pretty much "are you sure this is safe?"

Could something similar apply here?
